src/Components/Endpoints/src/Discovery/RazorComponentApplication.cs

```diff
@@ -1,26 +1,64 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using Microsoft.AspNetCore.Components.Web;
+
 namespace Microsoft.AspNetCore.Components;
 
 /// <summary>
 /// The definition of a component based application.
 /// </summary>
+[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
 public class RazorComponentApplication
 {
-    private readonly PageCollection _pageCollection;
+    private readonly PageComponentInfo[] _pages;
+    private readonly ComponentInfo[] _components;
 
-    // TODO: we define the concepts explicitly (like the collection of pages)
-    // In the future we need to decide if we want to do generalize this concept and use
-    // something "generic" like a list of "features".
-    internal RazorComponentApplication(PageCollection pagesCollection)
+    internal RazorComponentApplication(
+        PageComponentInfo[] pageCollection,
+        ComponentInfo[] componentCollection)
     {
-        _pageCollection = pagesCollection;
+        _pages = pageCollection;
+        _components = componentCollection;
     }
 
     /// <summary>
-    /// Gets the list of <see cref="PageDefinition"/> associated with the application.
+    /// Gets the list of <see cref="PageComponentInfo"/> associated with the application.
     /// </summary>
     /// <returns>The list of pages.</returns>
-    public IEnumerable<PageDefinition> Pages => _pageCollection.Pages;
+    public IReadOnlyList<PageComponentInfo> Pages => _pages;
+
+    /// <summary>
+    /// Gets the list of <see cref="ComponentInfo"/> associated with the application.
+    /// </summary>
+    public IReadOnlyList<ComponentInfo> Components => _components;
+
+    internal IEnumerable<IComponentRenderMode> ResolveRenderModes()
+    {
+        var set = new HashSet<IComponentRenderMode>();
+        for (var i = 0; i < Components.Count; i++)
+        {
+            var component = Components[i];
+            if (component.RenderMode is ServerRenderMode)
+            {
+                set.Add(RenderMode.Server);
+            }
+            if (component.RenderMode is WebAssemblyRenderMode)
+            {
+                set.Add(RenderMode.WebAssembly);
+            }
+            if (component.RenderMode is AutoRenderMode)
+            {
+                set.Add(RenderMode.Auto);
```

Comment:
Would it work to replace this with:

```suggestion
                set.Add(RenderMode.Server);
                set.Add(RenderMode.WebAssembly);
```

Reply:
I don't think so. Auto means it can render on both, not that it has to. If we were to do this, but only have webassembly enabled. is normal for auto to just coalesce into webassembly. It would be wrong to require the app to register server too, wouldn't it?

Reply:
OK, I think I had the wrong idea about what this is about. Would I have the right idea to suggest this method could be renamed to `GetRenderModesRequiredByDiscoveredComponents`?

And if so, what does it mean to say that auto is "required"? Is it equivalent to not saying that anything is required, since it doesn't force the developer to enable any specific configuration?

Reply:
Auto means whatever is available, if only server is available, uses server, if only wasm is available, uses wasm. If neither of them is configured, it throws to let you know, that you are either missing the AddServer/Webassembly components call or if you want to force static rendering, call `SetStaticRenderMode`